aptos-move/move-examples/sigma_protocols/sources/homomorphism.move

```diff
@@ -0,0 +1,414 @@
+///                                                                     (via Fiat-Shamir on $(X, A)$ and a domain-separator)
+///                                  6. send challenge $e$
+///                            <-------------------------------
+///
+/// 7. Compute response $\sigma = \alpha + e \cdot w$
+///
+///                               8. send response $\sigma$
+///                            ------------------------------->
+///
+///                                                                  9. Check $\psi(\sigma) = A + e f(X)$
+module sigma_protocols::homomorphism {
+    use std::bcs;
+    use aptos_std::ristretto255::{RistrettoPoint, Scalar, scalar_one, scalar_mul, new_scalar_uniform_from_64_bytes,
+        CompressedRistretto, point_identity,
+        multi_scalar_mul, point_equals
+    };
+    use std::error;
+    use aptos_std::aptos_hash::sha2_512;
+    use sigma_protocols::proof::Proof;
+    use sigma_protocols::secret_witness::SecretWitness;
+    use sigma_protocols::public_statement::PublicStatement;
+    use sigma_protocols::representation_vec::RepresentationVec;
+    use sigma_protocols::utils::{compress_points, points_clone, neg_scalars};
+
+    #[test_only]
+    use sigma_protocols::proof::{new_proof, empty_proof};
+    #[test_only]
+    use sigma_protocols::secret_witness::random_witness;
+    #[test_only]
+    use sigma_protocols::utils::{add_vec_points, add_vec_scalars, equal_vec_points, mul_points, mul_scalars};
+
+    //
+    // Error codes
+    //
+
+    /// The length of the `A` field in `Proof` did NOT match the homomorphism's output length
+    const E_PROOF_COMMITMENT_WRONG_LEN: u64 = 1;
+    /// The length of the `A` field in `Proof` should NOT be zero
+    const E_PROOF_COMMITMENT_EMPTY: u64 = 2;
+    /// One of our internal invariants was broken. There is likely a logical error in the code.
+    const E_INTERNAL_INVARIANT_FAILED: u64 = 3;
+
+    //
+    // Test-only error codes
+    //
+
+    #[test_only]
+    /// The slow verification of the sigma protocol proof failed (instead of succeeding) in one of the tests.
+    const E_SLOW_VERIFICATION_FAILED: u64 = 4;
+    #[test_only]
+    /// The fast verification of the sigma protocol proof failed (instead of succeeding) in one of the tests.
+    const E_FAST_VERIFICATION_FAILED: u64 = 5;
+
+    /// Unfortunately, we cannot directly use the `PublicStatement` struct here because its `vector<RistrettoPoint>`
+    /// will not serialize correctly via `bcs::to_bytes`, since a `RistrettoPoint` stores a Move VM "handle" rather than
+    /// an actual point.
+    struct FiatShamirInputs has drop {
+        dst: vector<u8>,
+        k: u64,
+        stmt_X: vector<CompressedRistretto>,
+        stmt_x: vector<Scalar>,
+        A: vector<CompressedRistretto>,
+    }
+
+    /// Returns the Sigma protocol challenge $e$ and $1,\beta,\beta^2,\ldots, \beta^{m-1}$
+    public fun fiat_shamir(
+        dst: vector<u8>,
```

Comment:
Domain separator is 3 things (from talking to Michele Orru):
 - Need a proof system identifier too (e.g., "Sigma protocol", chosen by the protocol)
 - Relation/homomorphism identifier
 - Session identifier (chosen by user; specifies the proof where this is valid; e.g., "Alice (0x1) is paying Bob (0x2) at time $T$")
     + Motivation prevent replay attacks (e.g., PoK of SK) or cross-protocol attacks